U/jrbogart/config reorg

JoanneBogart opened the pull request:
Reorganize config file so that almost everything is stored per object type. The top level file only includes schema version and the information necessary to locate the config and data directory and a mapping of object types to the yaml files containing details about them.  Since catalog files for different object types are in general made at different times, using different run options and possibly different versions of the code, provenance is stored per object type. Cosmology (for galaxy-like types) and SED information is also now stored within the per-object block.

The old object types used to represent components of galaxies (e.g. `bulge_basic`) are no longer full-fledged object types. Their information is in the parent file.

The creation code handles writing of the top-level config file and the file for the object type data being generated.  If the top-level file already exists because other object type data have already been created, the creation code will update the top file with an entry for the new object type. Entries for data not handled by the creation code (e.g. gaia stars; snana) must be added by hand. Examples of suitable per-object configs may be found under `skycatalogs/data/cfg_templates`

The SkyCatalog class can handle old-style configs as well as those in the new (schema version 1.3.0) format.

Some notes on details of and motivation for this update can be found [here](https://docs.google.com/presentation/d/1iNlQTO-gn8ozTkGHVR41U796EOr2VQKn1nmr6P6loec/edit?usp=sharing).

jchiang87 reviewed (changes requested):
I've made a few suggestions, but after getting part way through the changes, it might be better to just make a couple of general comments:

- There's a lot of code to make things backwards compatible.  For both near and long term maintainability, it might be preferable not to support backwards compatibility and instead provide scripts to convert existing catalogs to the new format.
- Instead of having separate utility functions for writing the provenance for each object type, I'm wondering if each BaseObject subclass could implement its own member function to write provenance.  I think that approach would avoid the code complexity that comes with the different if-elif branches in `config_utils.py` and would make it easier to add new object types in the future.

JoanneBogart commented on skycatalogs/utils/config_utils.py:
```diff
@@ -164,41 +234,74 @@ def __getitem__(self, k):
+            old_style = self._schema_pre_130
+            if not old_style:
+                if not object_type:
+                    if 'galaxy' in self._cfg['object_types']:
+                        object_type = 'galaxy'
+                    elif 'diffsky_galaxy' in self._cfg['object_types']:
+                        object_type = 'diffsky_galaxy'
```
There are no AGN currently, of course.   For snana-style SNe I don't see any mention of cosmology in the code. If it's needed it should probably be included in the snana section of the config (ideally to be supplied by the time domain people along with the data).   Or else there should be something connecting snana to the galaxy type.

I agree something needs to be done about multiple galaxy types - either support them or consistently disallow.

JoanneBogart commented on skycatalogs/utils/config_utils.py:
```diff
@@ -386,42 +446,221 @@ def assemble_file_metadata(pkg_root, inputs=None, run_options=None,
+          * it doesn't already exist     or
+          * we're allowed to overwrite
+        '''
+        if not self._overwrite:
+            try:
+                with open(outpath, mode='x') as f:
+                    yaml.dump(input_dict, f)
+            except FileExistsError:
+                txt = 'write_yaml: Will not overwrite pre-existing config'
+                self._logger.warning(txt + outpath)
+                return
+        else:
+            return self.update_yaml(input_dict, outpath)
 
```
agreed.

JoanneBogart commented on skycatalogs/utils/config_utils.py:
```diff
@@ -386,42 +446,221 @@ def assemble_file_metadata(pkg_root, inputs=None, run_options=None,
+            except FileExistsError:
+                txt = 'write_yaml: Will not overwrite pre-existing config'
+                self._logger.warning(txt + outpath)
+                return
```
It isn't used currently, but returning the path written seemed like it _might_ be useful.   In case of the exception nothing is written so `None` is returned.

jchiang87 replied:
Depending on how a return value would be used, I would nominally expect that
```
try:
    with open(outpath, mode='x') as f:
        yaml.dump(input_dict, f)
except FileExistsError:
    txt = 'write_yaml: Will not overwrite pre-existing config'
    self._logger.warning(txt + outpath)

return outpath
```
has the desired behavior, since in both cases (i.e., with and without the exception), the path to the yaml file with the specified config is returned.

JoanneBogart replied:
Returning None is an indication (available programmatically, not just in a log message) that nothing got written

jchiang87 replied:
Sure, but since nothing is using the return values, it's not clear that is or is not the right thing.

JoanneBogart commented on skycatalogs/utils/config_utils.py:
```diff
@@ -164,41 +234,74 @@ def __getitem__(self, k):
+                    else:
+                        return None
+                return self._cfg['object_types'][object_type]['Cosmology']
+        if old_style:
+            return self._cfg['Cosmology']
```
Originally `self._schema_pre_130` wasn't pre-computed so more complicated logic was needed here.   I neglected to simply here after that was changed.  Will fix.

JoanneBogart commented on skycatalogs/utils/config_utils.py:
```diff
@@ -164,41 +234,74 @@ def __getitem__(self, k):
+        if self._schema_version:
+            # cmps = self._cmps
+            # if (cmps[0] > 1) or (cmps[0] == 1 and cmps[1] > 2):
+            old_style = self._schema_pre_130
```
Nope

JoanneBogart commented:
Concerning your general comments -
Providing scripts to convert older configs to the current schema would no doubt be useful.   It would also take some work, significantly more than went into the code providing backward compatibility since there are plenty of differences in form which don't affect the operation of the API.   Also, when there is a change in structure like the one proposed here, there could be some awkwardness for people running an older version of the code and trying to access  a catalog with a new-style config.  The old code can't warn them what the problem is.  After a while presumably everyone would know to use a newer version of the code, but during the transition it could be frustrating. I propose that the code for this release should have backwards-compatibility more or less as currently implemented (taking your suggestions into account). Since I don't anticipate changing the location of `schema_version`, now a top-level key, ever again, I can put something in the code to check that version against what's in the code and put out a message if it's newer, so in the future there will be a warning if there is any kind of mismatch.
There should also be a script handling conversion when the schema changes, but I'd like to deal with that in a separate PR.
Moving the provenance-generating code to BaseObject and subclasses seems promising.  I'll try to get that in this PR if I don't run into any serious roadblocks.

jchiang87 commented:
An alternative to conversion scripts would be simply to make available modern versions of the catalogs in use now.  I doubt anybody aside from you and me have ever run `create_sc.py`.   My worry about the backwards compatibility code in there now is that it will be technical debt that never gets retired.  Also, because I had hoped it would go away, I didn't look as carefully as I would have had otherwise.

JoanneBogart commented:
I don't see how I can make use of BaseObject and subclasses as hoped.   There are no such objects lying around at the time the config fragments are being written.   I could make another class hierarchy for object _types_, but there would still need to be creation of an instance for each (or alternatively make everything static) and a mapping of  name to corresponding object-type object.  It wouldn't be any less code than is there now but maybe more maintainable.

JoanneBogart commented on skycatalogs/utils/config_utils.py:
```diff
@@ -386,42 +446,221 @@ def assemble_file_metadata(pkg_root, inputs=None, run_options=None,
 
 
-def get_file_metadata(fpath, key='provenance'):
+class ConfigWriter():
```
I've moved the code to create fragments to little per-object-type classes which subclass BaseConfigFragment.     The subclasses are within the xxx_object.py files.

jchiang87 commented:
One thing that I wanted to note is that the version info for the [throughputs package](https://github.com/lsst/throughputs/blob/1.9/baseline/version_info) is recorded in the `baseline` subfolder, e.g., for the descssim location at NERSC, there is
```
[login33 w_2024_22] pwd -P
/global/cfs/cdirs/descssim/imSim/lsst/data/throughputs_2023_09_07
[login33 w_2024_22] cat baseline/version_info 
1.9
[login33 w_2024_22] 
```
So that info could read from that location and added to the provenance output, presumably associated with the flux files.

JoanneBogart commented:
Thanks for the tip about throughputs version.  As I mentioned at the Thursday zoom, the plan was to include this information if possible.  I had seen those version numbers in the package but didn't know which, if any, was suitable for this purpose.

JoanneBogart commented:
I've added the throughputs version to the per-file metadata as suggested.  I've also added a little script to dump the provenance and updated sso file creation to include metadata (I forgot about sso in the first round).
The PR is ready for re-review.

jchiang87 commented on skycatalogs/objects/galaxy_object.py:
```diff
@@ -161,3 +162,20 @@ def get_instcat_entry(self, band='r', component=None):
+                 area_partition=None, data_file_type=None):
+        super().__init__(prov, object_type_name='galaxy')
+
+        self._opt_dict = {'area_partition': area_partition,
+                          'data_file_type': data_file_type}
+        self._cosmology = cosmology
+        self._tophat_bins = tophat_bins
+
```
Similar to my comment in `DiffskyConfigFragment`, why not just add `Cosmology` and `tophat` entries to `self._opt_dict`:
```
self._opt_dict = {'area_partition': area_partition,
                  'data_file_type': data_file_type,
                  'Cosmology': cosmology,
                  'tophat': {'bins': topat_bins}}
```
Along with the suggestion for `DiffskyConfigFragment`, this would enable all subclasses to use the same `make_fragment` implementation in the base class.

JoanneBogart replied:
I can add the `'Cosmology'` entry to `self._opt_dict` but not `'tophat'` because its value in the template is dict with a couple keys other than `'bins'` which would get lost with the update.

jchiang87 commented on skycatalogs/objects/star_object.py:
```diff
@@ -1,10 +1,11 @@
 import os
-import numpy as np
+# import numpy as np
```
Delete?

jchiang87 commented on skycatalogs/catalog_creator.py:
```diff
@@ -25,6 +24,9 @@
 from .objects.base_object import ROMAN_BANDS
+from .objects.star_object import StarConfigFragment
+from .objects.galaxy_object import GalaxyConfigFragment
+from .objects.diffsky_object import DiffskyConfigFragment
```
`SnanaConfigFragment` and `GaiaConfigFragment` aren't used anywhere?

JoanneBogart replied:
No, that's because skyCatalogs is not responsible for making those catalogs.   Those routines are provided in case someone (or some thing) that does produce them wants to also produce a suitable yaml fragment.

jchiang87 commented on skycatalogs/utils/config_utils.py:
```diff
@@ -384,44 +434,194 @@ def assemble_file_metadata(pkg_root, inputs=None, run_options=None,
-    object_types: iterable of (pointsource) object type names
+    inpath     string            path to file
+    silent     boolean           if file not found and silent is true,
+                                 return None.  Else raise exception
```
Please add a docstring entry for `resolve_include`.

jchiang87 reviewed (changes requested):
Thanks, the ConfigFragment implemetation looks much cleaner.  I do have some suggestions for simplifying/clarifying further.

Regarding making sure either only diffsky or cosmoDC2 galaxies are used, but not both, or alternatively, adding a mechanism to ensure cosmologies are consistent, can we add an issue to address this so that it doesn't get dropped entirely?  An issue to add cosmology info for SNe would also be good.

JoanneBogart commented on skycatalogs/objects/base_object.py:
```diff
@@ -70,7 +74,7 @@ def load_lsst_bandpasses():
         lsst_bandpasses[band] = bp
 
-    return lsst_bandpasses
+    return lsst_bandpasses, version
```
True.  load_lsst_bandpasses could instead store the version as well as the bandpasses in a separate global variable and there could be a separate get_lsst_bandpasses_version routine.  It would be neater to do this with a class rather than adding more global variables, but that also would break the interface.

jchiang87 replied:
I suggest an internal version of the function in addition to the public interface.  So change the name of the current `load_lsst_bandpasses` to `_load_lsst_bandpasses`, and make a new public function that just returns the dict, so something along the lines of
```
def load_lsst_bandpasses():
    return _load_lsst_bandpasses()[0]
```

JoanneBogart commented on skycatalogs/objects/base_object.py:
```diff
@@ -51,6 +50,11 @@ def load_lsst_bandpasses():
+        with open(os.path.join(bp_dir, 'version_info')) as f:
+            version = f.read().strip()
+    else:
+        version = 'galsim_builtin'
```
Concerning throughputs in `$HOME` - that code has been there forever (well, at least 2 years). I don't mind getting rid of it once the installation procedure is well-documented, which is a work in progress. See https://github.com/LSSTDESC/skyCatalogs/tree/u/jrbogart/skycat_doc.
I would rather not get rid of the galsim_builtin option (all there is currently for Roman) but there should be a warning; I'll add that.

jchiang87 replied:
Does anybody use the `$HOME` code?   I've always wondered why it's there.  afaik, that option isn't documented at all so I don't see why it even exists.

JoanneBogart replied:
I have no idea whether anyone uses it.  It's not documented in skyCatalogs documentation, but neither is anything else about finding throughputs or SEDs.

jchiang87 replied:
But we definitely know that the de facto standard functionality associated with throughputs and SEDs as implemented is used.  So, if it's possible to clean up this special case that no one uses, which has a preferred alternative, and which makes the code more complicated than it ultimately needs to be, I think we should do that.

JoanneBogart replied:
Although I don't know of anyone who uses the `$HOME` code, I can't be sure that no one does.   I'll get rid of it after the preferred method is documented.

JoanneBogart commented on skycatalogs/objects/base_config_fragment.py:
```diff
@@ -0,0 +1,61 @@
+        template_path = os.path.join(_TEMPLATE_DIR, self._template_name)
+        with open(template_path, 'r') as f:
+            data = yaml.load(f, Loader=yaml.SafeLoader)
+        if self._opt_dict:
+            opt = self._opt_dict
+            other = dict()
+            for key in opt:
+                if opt[key] is not None:
+                    other[key] = opt[key]
+            if len(other.keys()) > 0:
+                data.update(other)
```
I don't think I can just use `update`, at least not as the rest of the code stands. `self._opt_dict` will usually include keys like
```{'area_partition': None, 'data_file_type': None}```
The `None` should not override what's already there. So each of the subclass `__init__` routines would need to exclude those things from `self._opt_dict`.  I chose to do it in `generic_create` instead.

jchiang87 replied:
From what I can tell, all of the subclasses have `area_partition` and `data_file_type` as keyword options in their `__init__` functions, so it seems those could just be moved to the base class and the logic to exclude `None` put there.
